src/main/scala/analysis/IrreducibleLoops.scala

```diff
@@ -0,0 +1,389 @@
+import intrusivelist.{IntrusiveList}
+import util.Logger
+
+import scala.collection.mutable
+
+private def label(p: CFGPosition) = {
+p match {
+    case b: Block =>  "block." + b.label
+    case p: Procedure =>  "proc." + p.name
+    case  c  =>  "cmd." + c.hashCode().toString()
+}
+}
+
+
+
+/*
+ * Loop Identification
+ *
+ * 
+ */
+
+/* A connection between to IL nodes, purely for the representation of loops in
+ *
+ */
+class LoopEdge(val from: CFGPosition, val to: CFGPosition):
+    override def toString: String = s"(${label(from)}, ${label(to)})"
+    override def equals(n: Any): Boolean = n match {
+        case edge: LoopEdge => edge.from == from && edge.to == to;
+        case _ => false
+    }
+
+/* A loop is a subgraph <G_l, E_l> of a CFG <G, E>
+ *
+ */
+class Loop(var header: CFGPosition):
+    val reentries: mutable.Set[LoopEdge] = mutable.Set[LoopEdge]()    // Edges to loop from outside that are not to the header
+    val backEdges: mutable.Set[LoopEdge] = mutable.Set[LoopEdge]()    // Edges from inside loop to the header
+    val entryEdges: mutable.Set[LoopEdge] = mutable.Set[LoopEdge]()   // Edges into the header node
+
+    val nodes: mutable.Set[CFGPosition] = mutable.Set[CFGPosition]()    // G_l
+    val edges: mutable.Set[LoopEdge] = mutable.Set[LoopEdge]()          // G_e
+    var reducible: Boolean = true                                       // Assume reducible by default
+
+    def addEdge(edge: LoopEdge) = {
+        nodes += edge.from
+        nodes += edge.to
+        edges += edge
+    }
+
+    def name: String = label(header)
+
+    override def toString() = {
+        s"Header: ${label(header)}, Body: ${edges}"
+    }
+
+/* Loop detection and classification with respect to being reducible or irreducible. Implements the algorithm
+ *  described by Wei in `A New Algorithm for Identifying Loops in Decompilation` (LNCS 4632 pp 170-183)
+ */
+class LoopDetector(cfg: Program):
+
+    // Header -> Loop
+    val loops: mutable.HashMap[CFGPosition, Loop] = mutable.HashMap[CFGPosition, Loop]()
+    val headers: mutable.Set[CFGPosition] = mutable.Set[CFGPosition]()
+
+    // Algorithm helpers
+    val visitedNodes: mutable.Set[CFGPosition] = mutable.Set[CFGPosition]()
+    val nodeDFSPpos: mutable.HashMap[CFGPosition, Int] = mutable.HashMap[CFGPosition, Int]()
+    val iloopHeaders: mutable.HashMap[CFGPosition, CFGPosition] = mutable.HashMap[CFGPosition, CFGPosition]()
+    val edgeStack: mutable.Stack[LoopEdge] = mutable.Stack[LoopEdge]()
+
+    /* 
+     * Returns the set of irreducible loops in the program. 
+     *
+     */
+    def irreducible_loops(): Set[Loop] = {
+        val irreducibleLoops: Set[Loop] = loops.values.filter(l => !l.reducible).toSet;
+
+        if (irreducibleLoops.isEmpty) {
+            return Set[Loop]();
+        }
+
+        val wantedLoops: mutable.Set[Loop] = mutable.Set[Loop]();
+        val irrHeaders: Set[CFGPosition] = irreducibleLoops.map(l => l.header);
```

Comment:
What are these lines for?

Reply:
@ailrst do you have any idea what the idea was here?

Reply:
Don't know, as far as I can see they aren't used?